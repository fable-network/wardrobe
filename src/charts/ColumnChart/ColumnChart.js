import React from 'react';
import PropTypes from 'prop-types';
import styled, { css, withTheme } from 'styled-components';
import { theme as defaultTheme } from '../../theme';
import HighChart from '../HighChart';

const highlightedCss = css`
  .highcharts-point.highcharts-color-0 {
    fill: ${p => p.theme.grey06};

    &.highcharts-point-hover {
      fill: ${p => p.theme.backgroundInfo};
      fill-opacity: 1;
    }
  }
`;

const Wrapper = styled.div`
  position: relative;

  .ft-wardrobe-column-chart {
    .highcharts-point.highcharts-color-0 {
      &,
      &.highcharts-point-hover {
        fill: ${p => p.theme.backgroundInfo};
        fill-opacity: 1;
      }
    }
    ${p => p.highlighted && highlightedCss};

    .highcharts-yaxis-grid {
      display: none;
    }
    .highcharts-xaxis .highcharts-axis-line {
      stroke: ${p => p.theme.primaryActive};
      stroke-width: 2px;
    }
    div.highcharts-xaxis-labels {
      display: flex;
      font-size: ${p => p.theme.fontSizeSmall};
      line-height: ${p => p.theme.lineHeightSmall};
      font-weight: 400;
      color: ${p => p.theme.grey01};
      text-align: center;
    }
    .highcharts-credits {
      display: none;
    }
    .ft-label-block {
      display: block;
      text-align: center;
      box-sizing: border-box;

      .ft-label.ft-label-value {
        display: block;
        font-size: ${p => p.theme.fontSizeLarge};
        line-height: ${p => p.theme.lineHeightControlLarge};
        font-weight: ${p => p.theme.fontWeightBold};
        color: ${p => p.theme.grey01};
      }
      .ft-label.ft-label-caption {
        font-size: ${p => p.theme.fontSizeSmall};
        line-height: ${p => p.theme.lineHeightControlSmall};
        font-weight: ${p => p.theme.fontWeightNormal};
        color: ${p => p.theme.grey03};
      }

      &.ft-label-block-disabled {
        .ft-label.ft-label-value {
          color: ${p => p.theme.grey04};
        }
        .ft-label.ft-label-caption {
          color: ${p => p.theme.grey04};
        }
      }
    }
    .highcharts-subtitle {
      color: ${p => p.theme.grey01};
      font-size: ${p => p.theme.fontSizeSmall};
      line-height: ${p => p.theme.lineHeightSmall};
      font-weight: ${p => p.theme.fontWeightBold};
    }
  }
`;

class ColumnChart extends React.PureComponent {
  state = { highlightIndex: null };

  handleMouseOver = ({ target }) => {
    this.setState({ highlightIndex: target.index });
  };

  handleMouseOut = () => {
    this.setState({ highlightIndex: null });
  };

  renderLabel = ({ value, pos }) => (pos < this.props.data.length ? value : '');

  renderValue = (value, total) => {
    const { caption } = this.props;
    if (value === undefined) return '';
    const percentage = (100 * value.y) / total;
    const captionHtml = `<span class="ft-label ft-label-caption">
      ${typeof caption === 'function' && caption({ y: value.y, percentage })}
    </span>`;
    return `
      <span class="ft-label-block">
        <span class="ft-label ft-label-value">${value.y}</span>
        ${captionHtml}
      </span>
    `;
  };

  render() {
    const { data, title, subtitle, allowUpdate, theme = defaultTheme, height = null } = this.props;
    if (!data) {
      return null;
    }
    const { highlightIndex } = this.state;
    const { renderValue, renderLabel } = this;
    const total = data.reduce((sum, item) => sum + item.y, 0);
    const options = {
      title: title && {
        text: title,
      },
      subtitle: subtitle && {
        text: subtitle,
        align: 'right',
        verticalAlign: 'top',
      },
      chart: {
        type: 'column',
        backgroundColor: theme.white,
        styledMode: true,
        groupPadding: 0,
        height,
      },
      plotOptions: {
        series: {
          groupPadding: 0,
          events: {
            mouseOver: this.handleMouseOver,
            mouseOut: this.handleMouseOut,
          },
        },
        column: {
          borderRadius: 4,
          pointPadding: 0.01,
          groupPadding: 0,
          maxPointWidth: 230,
          dataLabels: {
            enabled: true,
            color: 'black',
            useHTML: true,
            formatter: function formatter() {
              return renderValue(this, total);
            },
          },
        },
      },
      xAxis: {
        // Min and minRange to make it generate extra void data points to align the only col left
        minRange: 3,
        min: 0,
        // It requires categories in advance to plan for space allocation
        categories: data.map(point => point.name),
        gridLineWidth: 0,
        tickWidth: 0,
        offset: 4,
        lineWidth: 0,
        labels: {
          useHTML: true,
          formatter: function formatter() {
            return renderLabel(this);
          },
        },
      },
      yAxis: {
        title: false,
        labels: { enabled: false },
        gridLineWidth: 0,
      },
      legend: {
        enabled: false,
      },
      tooltip: {
        enabled: false,
        outside: false, // Don't change unless you want to re-do the whole tooltip styling
        useHTML: true,
        shadow: true,
        backgroundColor: theme.white,
        borderColor: theme.grey04,
        borderRadius: 0,
        padding: 8,
      },
      series: [{ data }],
    };
    return (
      <Wrapper highlighted={highlightIndex !== null}>
        <HighChart
          options={options}
          className="ft-wardrobe-column-chart"
          allowChartUpdate={allowUpdate}
        />
      </Wrapper>
    );
  }
}

ColumnChart.defaultProps = {
  allowUpdate: false,
  caption: ({ y }) => y,
};

ColumnChart.propTypes = {
  title: PropTypes.string,
  subtitle: PropTypes.string,
  data: PropTypes.arrayOf(
    PropTypes.shape({
      y: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
    })
  ).isRequired,
  caption: PropTypes.func,
  height: PropTypes.number,
  /**
   * Allows HighCharts chart update. Use it if you plan to provide new data.
   */
  allowUpdate: PropTypes.bool,
  // from withTheme
  theme: PropTypes.object.isRequired,
};

export default withTheme(ColumnChart);
