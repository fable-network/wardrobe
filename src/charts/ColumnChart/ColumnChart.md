A column chart for a few data points. Example usage: delivery window distribution for an order.

```jsx
<ColumnChart
  title="Distribution of delivery"
  data={require('../fixtures/column-chart.js').default}
  caption={({ y }) => (y === 1 ? 'piece' : 'pieces')}
/>
```

With one data point:

```jsx
<ColumnChart
  title="Distribution of delivery"
  data={require('../fixtures/column-chart-one.js').default}
  caption={({ y }) => (y === 1 ? 'piece' : 'pieces')}
/>
```

With a lot of data points:

```jsx 
<ColumnChart
  title="Distribution of delivery"
  data={require('../fixtures/column-chart-many.js').default}
  caption={({ y }) => (y === 1 ? 'piece' : 'pieces')}
/>
```

(Morale: don't use it for a lot of data points)
