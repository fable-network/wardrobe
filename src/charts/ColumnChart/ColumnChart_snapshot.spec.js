/* eslint-disable */
// This file was automatically generated
import React from 'react';
import { shallow } from 'enzyme';
import ColumnChart from './ColumnChart';

describe('ColumnChart Component', () => {
  describe('snapshots', () => {
    it('should match snapshot', () => {
    
      const wrapper1 = shallow(
        <ColumnChart
          title="Distribution of delivery"
          data={require('../fixtures/column-chart.js').default}
          caption={({ y }) => (y === 1 ? 'piece' : 'pieces')}
        />
      );
      expect(wrapper1).toMatchSnapshot();

      const wrapper2 = shallow(
        <ColumnChart
          title="Distribution of delivery"
          data={require('../fixtures/column-chart-one.js').default}
          caption={({ y }) => (y === 1 ? 'piece' : 'pieces')}
        />
      );
      expect(wrapper2).toMatchSnapshot();

    });
  });
});
