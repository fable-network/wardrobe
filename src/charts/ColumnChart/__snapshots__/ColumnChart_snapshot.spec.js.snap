// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnChart Component snapshots should match snapshot 1`] = `
.c0 {
  position: relative;
}

.c0 .ft-wardrobe-column-chart .highcharts-point.highcharts-color-0,
.c0 .ft-wardrobe-column-chart .highcharts-point.highcharts-color-0.highcharts-point-hover {
  fill: #e7f1f7;
  fill-opacity: 1;
}

.c0 .ft-wardrobe-column-chart .highcharts-yaxis-grid {
  display: none;
}

.c0 .ft-wardrobe-column-chart .highcharts-xaxis .highcharts-axis-line {
  stroke: #034a76;
  stroke-width: 2px;
}

.c0 .ft-wardrobe-column-chart div.highcharts-xaxis-labels {
  font-size: 0.75rem;
  line-height: 1.3333333333333333;
  font-weight: 400;
  color: #2d3b43;
}

.c0 .ft-wardrobe-column-chart .highcharts-credits {
  display: none;
}

.c0 .ft-wardrobe-column-chart .ft-label-block {
  display: block;
  padding-top: 0.25rem;
  text-align: center;
  box-sizing: border-box;
}

.c0 .ft-wardrobe-column-chart .ft-label-block .ft-label.ft-label-value {
  display: block;
  font-size: 1.125rem;
  line-height: 1.3333333333333333;
  font-weight: 500;
  color: #2d3b43;
}

.c0 .ft-wardrobe-column-chart .ft-label-block .ft-label.ft-label-caption {
  font-size: 0.875rem;
  line-height: 1.1428571428571428;
  font-weight: 400;
  color: #949ea3;
}

.c0 .ft-wardrobe-column-chart .ft-label-block.ft-label-block-disabled .ft-label.ft-label-value {
  color: #c1c7cb;
}

.c0 .ft-wardrobe-column-chart .ft-label-block.ft-label-block-disabled .ft-label.ft-label-caption {
  color: #c1c7cb;
}

.c0 .ft-wardrobe-column-chart .highcharts-subtitle {
  color: #2d3b43;
  font-size: 0.875rem;
  line-height: 1.4285714285714286;
  font-weight: 500;
}

<WithTheme(ColumnChart)
  caption={[Function]}
  data={
    Array [
      Object {
        "name": "Immediate",
        "y": 231,
      },
      Object {
        "name": "Jan 15, 2019 — Jan 28, 2019",
        "y": 91,
      },
      Object {
        "name": "Feb 15, 2019 — Mar 28, 2019",
        "y": 321,
      },
      Object {
        "name": "Apr 15, 2019 — May 28, 2019",
        "y": 81,
      },
      Object {
        "name": "Jun 15, 2019 — Jul 28, 2019",
        "y": 263,
      },
    ]
  }
  subtitle="Total: 480 pieces"
  title="Distribution of delivery"
>
  <ColumnChart__Wrapper
    highlighted={false}
  >
    <div
      className="c0"
    >
      <high-charts />
    </div>
  </ColumnChart__Wrapper>
</WithTheme(ColumnChart)>
`;

exports[`ColumnChart Component snapshots should match snapshot 2`] = `
.c0 {
  position: relative;
}

.c0 .ft-wardrobe-column-chart .highcharts-point.highcharts-color-0,
.c0 .ft-wardrobe-column-chart .highcharts-point.highcharts-color-0.highcharts-point-hover {
  fill: #e7f1f7;
  fill-opacity: 1;
}

.c0 .ft-wardrobe-column-chart .highcharts-yaxis-grid {
  display: none;
}

.c0 .ft-wardrobe-column-chart .highcharts-xaxis .highcharts-axis-line {
  stroke: #034a76;
  stroke-width: 2px;
}

.c0 .ft-wardrobe-column-chart div.highcharts-xaxis-labels {
  font-size: 0.75rem;
  line-height: 1.3333333333333333;
  font-weight: 400;
  color: #2d3b43;
}

.c0 .ft-wardrobe-column-chart .highcharts-credits {
  display: none;
}

.c0 .ft-wardrobe-column-chart .ft-label-block {
  display: block;
  padding-top: 0.25rem;
  text-align: center;
  box-sizing: border-box;
}

.c0 .ft-wardrobe-column-chart .ft-label-block .ft-label.ft-label-value {
  display: block;
  font-size: 1.125rem;
  line-height: 1.3333333333333333;
  font-weight: 500;
  color: #2d3b43;
}

.c0 .ft-wardrobe-column-chart .ft-label-block .ft-label.ft-label-caption {
  font-size: 0.875rem;
  line-height: 1.1428571428571428;
  font-weight: 400;
  color: #949ea3;
}

.c0 .ft-wardrobe-column-chart .ft-label-block.ft-label-block-disabled .ft-label.ft-label-value {
  color: #c1c7cb;
}

.c0 .ft-wardrobe-column-chart .ft-label-block.ft-label-block-disabled .ft-label.ft-label-caption {
  color: #c1c7cb;
}

.c0 .ft-wardrobe-column-chart .highcharts-subtitle {
  color: #2d3b43;
  font-size: 0.875rem;
  line-height: 1.4285714285714286;
  font-weight: 500;
}

<WithTheme(ColumnChart)
  caption={[Function]}
  data={
    Array [
      Object {
        "name": "Immediate",
        "y": 42,
      },
    ]
  }
  height={300}
  subtitle="Total: 42 pieces"
  title="Distribution of delivery"
>
  <ColumnChart__Wrapper
    highlighted={false}
  >
    <div
      className="c0"
    >
      <high-charts />
    </div>
  </ColumnChart__Wrapper>
</WithTheme(ColumnChart)>
`;
